### DISCUSSION TITLE
WhoBPyT Architecture Plan

### DISCUSSION THREAD

**Discussion by JohnGriffiths** [GriffithsLab/whobpyt#26] [Ideas]
## WhoBPyT Architecture Planning

This thread will be used for discussions on two related things: 

a) the whobpyt code base architecture,  and 

b) the general development model

There is a general consensus that progress so far has been hampered by lack of clarity on these two points. 

The **location of the current consensus on the whobpyt architecture is [this file](https://github.com/GriffithsLab/whobpyt/blob/main/doc/about_whobyt/architecture.md)**

Please read this document. The proximal result of discussions in this thread should be updates to this file. 

Please submit PRs with updates to the document contents as we go.

**Comment by Andrew-Clappison** [comment]
Just as a start - Decisions to make on included features and other choices (which we need to decide is in/out of scope for base code V1 of WhoBPyT). This will influence architecture in terms of classes we need, and the expected input/output of these classes. 

Applicable to all:
- Implement Option of batching 
- GPU Support
- Optional Laplacian
- Boundary for fitted parameters
- Mean and variance of Priors
- Local Node Parameter Fitting
- What Objective Function Components
- "Sampling" in training function
- Boundary Conditions on State Variables
- fNIRS BOLD feature
- How to combine multiple objective function components
- Standard data structures/dimensions for input/output data
- Allowing use of other optimizers other than ADAM
- How we will represent simulated and empirical data internal to the class. 
- What output we want by default

Jansen-Rit
- Delays
- How connections will be implemented

RWW
- Staying with RWW Parameters or with g_IE, g_EE, etc...
- Time Changes
- Method to fit simulation in memory by having forward run to get initial conditions then running in GPU
- Choise of solution to RWW firing rate equations unstable 
- Whether we allow any additional tanh() to be optionally added
- Adding gammaI term and J_new term to RWW

Linear
- What options to give
